File: src/crewlet/sandbox/coding_agents/claude_code.py
"""``ClaudeCodeRunner`` — Claude Code headless inside an E2B sandbox.

Invokes the Claude Code CLI (preinstalled in E2B's ``claude`` template)
in headless JSON mode and maps its output onto :class:`CodingAgentResult`.
The detached file plumbing (start/poll/collect, the ask overlay) lives in
:class:`DetachedFileRunner`; this module supplies the ``claude -p``
command builder and the result-JSON parser — both pure + unit-pinned.
"""

from __future__ import annotations

import json
import shlex

from crewlet.sandbox.coding_agents._detached import (
    ASK_PATH,
    DONE_MARKER,
    ERR_PATH,
    PR_RE,
    RESULT_PATH,
    WORK_DIR,
    DetachedFileRunner,
)
from crewlet.sandbox.protocol import CodingAgentLLM, CodingAgentResult, RunLimits

# Re-exported for callers / tests that pin the sandbox artifact paths.
__all__ = [
    "ASK_PATH",
    "DONE_MARKER",
    "ERR_PATH",
    "RESULT_PATH",
    "WORK_DIR",
    "ClaudeCodeRunner",
    "build_claude_command",
    "parse_claude_result",
]


def build_claude_command(
    brief: str,
    *,
    model: str = "",
    limits: RunLimits | None = None,
    session_id: str = "",
    mcp_config_path: str = "",
) -> str:
    """Build the headless ``claude -p`` invocation.

    Pure + deterministic so the exact flags are unit-pinned. The brief is
    shell-quoted; budget caps are appended only when set so a minimal call
    stays minimal. Runs ``--permission-mode bypassPermissions``, so the agent
    may use every tool the scoped MCP servers expose — there is no per-tool
    allowlist.
    """
    parts = [
        "claude",
        "-p",
        shlex.quote(brief),
        "--output-format",
        "json",
        "--permission-mode",
        "bypassPermissions",
    ]
    if model:
        parts += ["--model", shlex.quote(model)]
    if session_id:
        parts += ["--resume", shlex.quote(session_id)]
    if limits is not None:
        if limits.max_turns > 0:
            parts += ["--max-turns", str(limits.max_turns)]
        if limits.max_budget_usd > 0:
            parts += ["--max-budget-usd", str(limits.max_budget_usd)]
    if mcp_config_path:
        parts += ["--mcp-config", shlex.quote(mcp_config_path), "--strict-mcp-config"]
    return " ".join(parts)


def parse_claude_result(stdout: str) -> CodingAgentResult:
    """Map Claude Code's ``--output-format json`` output onto a result.

    Tolerant: non-JSON / partial output yields a failed result carrying
    the raw text as ``error`` rather than raising — a coding agent that
    crashed should surface as "didn't deliver", not blow up the turn.
    """
    text = (stdout or "").strip()
    if not text:
        return CodingAgentResult(success=False, error="empty coding-agent output")
    try:
        obj = json.loads(text)
    except (json.JSONDecodeError, ValueError):
        # The CLI sometimes prints a banner before the JSON object; try
        # the last line, then give up gracefully.
        last = text.splitlines()[-1].strip()
        try:
            obj = json.loads(last)
        except (json.JSONDecodeError, ValueError):
            return CodingAgentResult(
                success=False, text=text[:2000], error="unparseable coding-agent output"
            )
    if not isinstance(obj, dict):
        return CodingAgentResult(success=False, error="unexpected output shape")

    result_text = str(obj.get("result", "") or "")
    usage = obj.get("usage") or {}
    input_tokens = int(usage.get("input_tokens", 0) or 0)
    output_tokens = int(usage.get("output_tokens", 0) or 0)
    cost = float(obj.get("total_cost_usd", 0.0) or 0.0)
    is_error = bool(obj.get("is_error", False))
    subtype_ok = obj.get("subtype", "success") == "success"
    success = subtype_ok and not is_error
    refs = PR_RE.findall(result_text)
    return CodingAgentResult(
        text=result_text,
        success=success,
        input_tokens=input_tokens,
        output_tokens=output_tokens,
        cost_usd=cost,
        session_id=str(obj.get("session_id", "") or ""),
        delivered_refs=refs,
        error="" if success else (str(obj.get("error", "")) or result_text[:500]),
    )


class ClaudeCodeRunner(DetachedFileRunner):
    """Runs Claude Code headless in a sandbox; returns a CodingAgentResult."""

    name = "claude-code"

    def _build_command(
        self,
        brief: str,
        limits: RunLimits,
        *,
        llm: CodingAgentLLM | None = None,
        mcp_config_path: str = "",
    ) -> str:
        # Claude Code takes the bare model id; its endpoint + key come from
        # the sandbox env (ANTHROPIC_BASE_URL / ANTHROPIC_API_KEY).
        return build_claude_command(
            brief,
            model=llm.model if llm else "",
            limits=limits,
            mcp_config_path=mcp_config_path,
        )

    def _parse_result(self, stdout: str) -> CodingAgentResult:
        return parse_claude_result(stdout)

File: internal/sandbox/codingagent/claudecode.go
package codingagent

import (
	"context"
	"encoding/json"
	"strconv"
	"strings"

	"github.com/crewlet/crewlet/internal/sandbox"
)

// ClaudeCodeName is this runner's config name.
const ClaudeCodeName = "claude-code"

// ClaudeCode drives Claude Code headless.
//
// It reaches its model through the run ENVIRONMENT rather than a config file,
// which is why WriteConfig here only ever renders MCP servers: the credential
// comes from providers.llm via the sandbox env, and duplicating it into a file
// inside the box would put the secret somewhere the agent can read it back.
type ClaudeCode struct{}

var _ CLI = ClaudeCode{}

// NewClaudeCode returns the runner.
func NewClaudeCode() *Runner { return New(ClaudeCode{}) }

// Name is the coding agent's key in the runner registry.
func (ClaudeCode) Name() string { return ClaudeCodeName }

// Command builds the headless `claude -p` invocation.
//
// Pure and deterministic so the exact flags can be pinned by a test. Budget
// caps are appended only when set, so a minimal call stays minimal.
//
// It runs with permissions bypassed, and that is a deliberate consequence of
// the scoping model rather than an oversight: the box IS the boundary, and
// there is no per-tool allowlist inside it. What the agent may reach is
// decided by which MCP servers were rendered into its config and which
// credentials the run env carries — both engine-side, both before the agent
// starts. A permission prompt would simply hang a headless run.
func (ClaudeCode) Command(req sandbox.RunRequest, _ Paths, configPath string) string {
	parts := []string{
		"claude", "-p", shellQuote(req.Brief),
		"--output-format", "json",
		"--permission-mode", "bypassPermissions",
	}
	if req.LLM != nil && req.LLM.Model != "" {
		parts = append(parts, "--model", shellQuote(req.LLM.Model))
	}
	if req.Limits.MaxTurns > 0 {
		parts = append(parts, "--max-turns", strconv.Itoa(req.Limits.MaxTurns))
	}
	if req.Limits.MaxBudgetUSD > 0 {
		parts = append(parts, "--max-budget-usd",
			strconv.FormatFloat(req.Limits.MaxBudgetUSD, 'f', -1, 64))
	}
	if configPath != "" {
		// --strict-mcp-config so the agent gets EXACTLY the scoped surface:
		// without it the CLI also loads whatever config the box's home
		// happens to carry, which on a reused box is the previous run's.
		parts = append(parts, "--mcp-config", shellQuote(configPath), "--strict-mcp-config")
	}
	return strings.Join(parts, " ")
}

// WriteConfig renders the scoped MCP surface, or nothing when there is none.
func (ClaudeCode) WriteConfig(ctx context.Context, box sandbox.Sandbox, req sandbox.RunRequest, paths Paths) (string, error) {
	if len(req.MCPServers) == 0 {
		return "", nil
	}
	servers := make(map[string]any, len(req.MCPServers))
	for name, s := range req.MCPServers {
		servers[name] = claudeCodeMCP(s)
	}
	blob, err := json.MarshalIndent(map[string]any{"mcpServers": servers}, "", "  ")
	if err != nil {
		return "", err
	}
	if err := box.WriteFile(ctx, paths.MCPConfig(), blob); err != nil {
		return "", err
	}
	return paths.MCPConfig(), nil
}

// claudeCodeMCP is one server in this CLI's own .mcp.json vocabulary.
//
// THE KEY NAMES ARE THIS RUNNER'S, which is why they live here: they are what
// `claude --mcp-config` reads, and nothing above this file should have to know
// them. internal/sandbox used to spell them out and then say in its own doc
// that they belonged to a runner.
//
// An empty `env` or `headers` is OMITTED rather than written as an empty
// object, which is what the CLI's own examples show and what this wrote
// before.
func claudeCodeMCP(s sandbox.MCPServer) map[string]any {
	if s.Transport == sandbox.TransportHTTP {
		out := map[string]any{"type": string(sandbox.TransportHTTP), "url": s.URL}
		if len(s.Headers) > 0 {
			out["headers"] = s.Headers
		}
		return out
	}
	out := map[string]any{"command": s.Command, "args": s.Args}
	if len(s.Env) > 0 {
		out["env"] = s.Env
	}
	return out
}

// Finished is false: this CLI exits cleanly, so the done marker is the signal.
func (ClaudeCode) Finished(string) bool { return false }

// Parse maps the CLI's JSON output onto a result.
//
// TOLERANT BY DESIGN: non-JSON or partial output yields a FAILED result
// carrying the raw text, never an error. A coding agent that crashed should
// surface as "did not deliver" and let the turn continue, not blow the turn up
// — the executor can still report what happened, which is more use to the
// requester than a failed turn.
func (ClaudeCode) Parse(stdout string) sandbox.Result {
	text := strings.TrimSpace(stdout)
	if text == "" {
		return sandbox.Result{Error: "the coding agent produced no output"}
	}
	obj, ok := decodeObject(text)
	if !ok {
		// TAILED, not head-cut. Unparseable output is the case where the
		// text IS the result — there is no structured field to fall back
		// to — and the useful part of it (the actual error, after the
		// banner and the warnings) is at the END, which is exactly what a
		// 2000-byte head cut discarded. Bounded because this is the CLI's
		// whole stdout and nothing upstream limits it; marked, so a reader
		// can tell a cut from a short run.
		return sandbox.Result{
			Text:  tail(text),
			Error: "the coding agent's output could not be parsed",
		}
	}

	resultText := stringField(obj, "result")
	// subtype names HOW the run ended ("success", "error_max_turns", …), and
	// is_error whether it failed. Both must be right: a run that hit its
	// turn cap reports no is_error but did not finish.
	success := stringOr(obj, "subtype", "success") == "success" && !boolField(obj, "is_error")

	res := sandbox.Result{
		Text:          resultText,
		Success:       success,
		SessionID:     stringField(obj, "session_id"),
		CostUSD:       floatField(obj, "total_cost_usd"),
		DeliveredRefs: prPattern.FindAllString(resultText, -1),
	}
	if usage, ok := obj["usage"].(map[string]any); ok {
		res.InputTokens = intField(usage, "input_tokens")
		res.OutputTokens = intField(usage, "output_tokens")
	}
	if !success {
		res.Error = stringField(obj, "error")
		if res.Error == "" {
			res.Error = resultText
		}
	}
	return res
}

// decodeObject reads a JSON object, falling back to the LAST line.
//
// The CLI sometimes prints a banner or a warning before its JSON, so a whole-
// text parse fails on output that is perfectly good — and the object is always
// last, because it is the thing the CLI prints when it is done.
func decodeObject(text string) (map[string]any, bool) {
	var obj map[string]any
	if json.Unmarshal([]byte(text), &obj) == nil && obj != nil {
		return obj, true
	}
	lines := strings.Split(text, "\n")
	last := strings.TrimSpace(lines[len(lines)-1])
	if last == "" {
		return nil, false
	}
	if json.Unmarshal([]byte(last), &obj) == nil && obj != nil {
		return obj, true
	}
	return nil, false
}

func stringField(obj map[string]any, key string) string {
	s, _ := obj[key].(string)
	return s
}

func stringOr(obj map[string]any, key, fallback string) string {
	if s, ok := obj[key].(string); ok {
		return s
	}
	return fallback
}

func boolField(obj map[string]any, key string) bool {
	b, _ := obj[key].(bool)
	return b
}

func floatField(obj map[string]any, key string) float64 {
	switch v := obj[key].(type) {
	case float64:
		return v
	case json.Number:
		f, _ := v.Float64()
		return f
	}
	return 0
}

func intField(obj map[string]any, key string) int {
	switch v := obj[key].(type) {
	case float64:
		return int(v)
	case json.Number:
		n, _ := v.Int64()
		return int(n)
	}
	return 0
}
